sdk/eventhub/azure-eventhubs/azure/eventhub/_producer_client.py

```diff
@@ -0,0 +1,170 @@
+            :start-after: [START create_eventhub_producer_client_sync]
+            :end-before: [END create_eventhub_producer_client_sync]
+            :language: python
+            :dedent: 4
+            :caption: Create a new instance of the EventHubProducerClient.
+
+    """
+    def __init__(self, host, event_hub_path, credential, **kwargs):
+        # type:(str, str, Union[EventHubSharedKeyCredential, EventHubSASTokenCredential, TokenCredential], Any) -> None
+        super(EventHubProducerClient, self).__init__(
+            host=host, event_hub_path=event_hub_path, credential=credential, **kwargs)
+        self._producers = []  # type: List[EventHubProducer]
+        self._client_lock = threading.Lock()
+        self._producers_locks = []  # type: List[threading.Lock]
+        self._max_message_size_on_link = 0
+
+    def _init_locks_for_producers(self):
+        if not self._producers:
+            with self._client_lock:
+                if not self._producers:
+                    num_of_producers = len(self.get_partition_ids()) + 1
+                    self._producers = [None] * num_of_producers
+                    for _ in range(num_of_producers):
+                        self._producers_locks.append(threading.Lock())
+
+    def send(self, event_data, **kwargs):
+        # type: (Union[EventData, EventDataBatch, Iterable[EventData]], Any) -> None
+        """
+        Sends an event data and blocks until acknowledgement is received or operation times out.
+
+        :param event_data: The event to be sent. It can be an EventData object, or iterable of EventData objects.
+            :class: `azure.eventhub.EventData`, `azure.eventhub.EventDataBatch`, `Iterator`, `Generator`, `list`
+        :keyword str partition_key: With the given partition_key, event data will land to
+         a particular partition of the Event Hub decided by the service.
+        :keyword str partition_id: The specific partition ID to send to. Default is None, in which case the service
+         will assign to all partitions using round-robin.
+        :keyword float timeout: The maximum wait time to send the event data.
+         If not specified, the default wait time specified when the producer was created will be used.
+        :rtype: None
+
+        .. admonition:: Example:
+
+            .. literalinclude:: ../samples/sync_samples/sample_code_eventhub.py
+                :start-after: [START eventhub_producer_client_send_sync]
+                :end-before: [END eventhub_producer_client_send_sync]
+                :language: python
+                :dedent: 4
+                :caption: Sends an event data and blocks until acknowledgement is received or operation times out.
+
+        """
+        partition_id = kwargs.pop("partition_id", None)
+
+        self._init_locks_for_producers()
+
+        producer_index = int(partition_id) if partition_id is not None else -1
+        if self._producers[producer_index] is None or\
```

Comment:
Using an index of -1 will get the last item in the list - is that the intention?

Reply:
See my other comment about perhaps using a dictionary instead of lists/indexes

Reply:
Non-blocking for now.